genesis/engine/solvers/fem_solver.py

```diff
@@ -1272,3 +1309,120 @@ def n_elements_max(self):
     @property
     def vol_scale(self):
         return self._vol_scale
+   
+
+    # ------------------------------------------------------------------------------------
+    # -------------------------------- vertex constraints --------------------------------
+    # ------------------------------------------------------------------------------------
+
+    def set_vertex_constraints(self, vertex_indices, target_positions, constraint_type="hard", stiffness=1e6):
+        """
+        Set vertex constraints for specified vertices.
+        
+        Parameters
+        ----------
+            vertex_indices : array_like
+                List of vertex indices to constrain
+            target_positions : array_like, shape (len(vertex_indices), 3)
+                List of target positions [x, y, z] for each vertex
+            constraint_type: str
+                A "hard" constraint directly set position and zero velocity.
+                A "soft" constraint uses a spring force to pull the vertex towards the target position.
+            stiffness : None | float, optional
+                Fixed position by default. Specify a spring stiffness for a soft constraint. (TODO: what's a reasonable range?)
+        """
+        if self._use_implicit_solver:
+            gs.logger.warning("Vertex constraints are not supported when using implicit solver. Constraint addition ignored.")
+            return
+ 
+        constraint_type_int = 0 if constraint_type == "hard" else 1
+        
+        vertex_indices_np = np.array(vertex_indices, dtype=np.int32)
+        target_positions_np = np.array(target_positions, dtype=np.float32)
+        
+        self._kernel_set_vertex_constraints(vertex_indices_np, target_positions_np, constraint_type_int, stiffness)
+
+    def update_constraint_targets(self, vertex_indices, new_target_positions):
+        """Update target positions for existing constraints."""
+        if not isinstance(vertex_indices, (list, tuple)):
```

Comment:
Must accept many other types for consistency with Rigid Solver, i.e. slice, range, torch.Tensor, np.ndarray...etc

Take inspiration from `_sanitize_1D_io_variables`